### ISSUE TITLE
Connection to metno broken after update to KF-5.92.0

### ISSUE THREAD

**Issue by martin-ms** [blackadderkate/weather-widget-2#89]
Yesterday I did a full system update and got the KDE Frameworks 5.92. Since then the weather-widget was no longer able to display data from metno, the circle spinning around for a few seconds and with the result of "-- N/A", while the connections to OWM are not affected. 

It taken some time & tests to narrow myself to the problem that it probably belongs to the "systemmonitor"-widget; I had one on the desktop to keep an eye on the disk usage.  If I remove that widget and logoff/logon the desktop, the weather-widget could display again data from metno, if I place again the "systemmonitor"-widget on the desktop, after logoff/logon I lost the connection to metno.  Just the presence of the widget is enough to prove this strange behaviour, no configuration needed. 

Another interesting thing is that the problem only occurs if you have an instance running in the system tray.  If there is a one & only task on the desktop, it works under all circumstances; if you have one in the systray, other instances on the desktop show the same problem. 

That happen with the current version of the weather widget as well as with that official supplied from the KDE Store.  Since both, the weather- and the systemmonitor-widget did not changed in the past days but KF-5.92, I suppose that changes in the KDE Frameworks causes the problem.

I am not deep enough into the topic to understand the interaction between these both widgets, so I can currently only give the information that it does not proper run on the KF-5.92 but don't know, where corrections needed; but I am afraid that sooner or later it will affect more users if distributors spread the latest release of the KDE Frameworks around.  I already compiled the weather-widget and the whole bunch of the plasma-workspace (which includes the systemmonitor-widget) against a KF-5.92-system, but with no progress.  It happen on two systems, this desktop computer and also a notebook.

I will also file a bugreport to the distributor, he might probably better confirm the problem in a KF-5.92-environment, but maybe you have also an idea how to solve or where else to report it.

**Comment by fkru** [comment]
> 
I am with KDE Frameworks 5.92 on openSUSE Tumbleweed since last week and I have also noticed yesterday that Met.no data were not updated for almost 12 hrs. This happens from time to time and I don't think this is due to the weather widget (or KDE) rather than their server.

**Comment by cfint** [comment]
Same issue here. I thought I was throttled by the metno API.

I can reproduce the original report. Remove system monitor widgets from tray. Restart plasmashell. The metno provider loads again.

The XMLHttpRequest is returning 403 forbidden.

**Comment by martin-ms** [comment]
Thank you for your response and the suggestions.

> * Edit the file:  **~/.local/share/plasma/plasmoids/org.kde.weatherWidget-2/contents/ui/main.qml**, and change line 109 to read **property bool debugLogging: true**

I made the changes and logged out and in, but there was no file **~/.xsession-errors**  written.  Where is the debug output expected?

> It would also be helpful if you could let me know what Distribution and Version of Linux you're using...

System is
```
Operating System: Arch Linux
KDE Plasma Version: 5.24.3
KDE Frameworks Version: 5.92.0
Qt Version: 5.15.3
Kernel Version: 5.16.15-arch1-1 (64-bit)
Graphics Platform: X11
```
The bugreport on https://bugs.archlinux.org/task/74182 has already been closed by the maintainer with the comment "This has nothing to do with packaging - please take it upstream", so we are on our own now.

**Comment by martin-ms** [comment]
>  This happens from time to time and I don't think this is due to the weather widget (or KDE) rather than their server.

**Can you please try these steps to find out what really happen:**

- get the "Weather Widget 2" from the official KDE Store, if not already done
- right click into the taskbar and add the widget. There are by default already three connections, and you will see the weather data for "Leicester, UK" from the metno sources, It's important to place it in the control panel, and not on the desktop!
- now place the "System monitor Sensor"-widget on the desktop; there is no need for further configuration, the simple presence is sufficient
- after signoff/signin you should see that the connections to the metno places are broken, indicated by a "-- N/A" but that on OWM is still alive
- remove the "System monitor Sensor"-widget from desktop and signoff/signin; connection to metno should work now again

It was also my first idea that the server was temporarly down, but if you follow the link at the bottom of the meteogram while the status is "N/A" you will get a connection to the website https://api.met.no and therefore it seems that the server respond correctly.

**Comment by blackadderkate** [comment]
@martin-ms

Archlinux have closed your report, because it is a bug in KDE or Qt.
I will file a bug for this on the KDE website as soon as I can, and post a comment on the KDE store to let people know there is an issue with new distributions...
Unfortunately I don't have the option to install the latest version of Archlinux or openSUSE to diagnose the problem myself.


If you can't find the .xsession-errors file, try the following:

In konsole:
- Set debuglogging to true as described before. 
- kill the plasmashell process.
- type the following : **plasmashell > /tmp/logfile 2>&1**

That should store the logs in /tmp/logfile 
Let it run for a minute or two, and then:

- Attach the /tmp/logfile to this bug report (scrubbing any personal data)...
- Set debuglogging back to the default as described before.
- kill plasmashell 
- restart plasmashell with: **plasmashell & **

**Comment by martin-ms** [comment]
> Archlinux have closed your report, because it is a bug in KDE or Qt. I will file a bug for this on the KDE website as soon as I can, and post a comment on the KDE store to let people know there is an issue with new distributions... 

Please tell us the link if you done that I can join it.  I would also file a bugreport on KDE but I still don't know the component who is responsible for the problem.  And I don't like to revert package by package to a previous version until I found it. 

> * type the following : **plasmashell > /tmp/logfile 2>&1**

I did it, but sorry to tell, that I got only a file with 0-length without any output.

**Comment by solbu** [comment]
When I needed to fetch the logging for Kate when debugging my issue in #81
I found my debugging logs in the syslog, so i ran `«jourcnalctl -f»` in a terminal, not sure if you need to be root/using sudo to see them.

Btw, the proper way to start KDE5/Plasma is `«kstart5 plasmashell»`
Whenever I need to restart my plasma environment, I run this, in a new terminal, which I _close_ afterwards, to stop the login messages from showing up in the terminal:
`«kquitapp5 plasmashell && kstart5 plasmashell»`

**Comment by blackadderkate** [comment]
I have filed a bug report on the KDE bug-tracking site.
[https://bugs.kde.org/show_bug.cgi?id=451790](https://bugs.kde.org/show_bug.cgi?id=451790)

**Comment by blackadderkate** [comment]
> > Archlinux have closed your report, because it is a bug in KDE or Qt. I will file a bug for this on the KDE website as soon as I can, and post a comment on the KDE store to let people know there is an issue with new distributions...
> 
> Please tell us the link if you done that I can join it. I would also file a bugreport on KDE but I still don't know the component who is responsible for the problem. And I don't like to revert package by package to a previous version until I found it.
> 
> > * type the following : **plasmashell > /tmp/logfile 2>&1**
> 
> I did it, but sorry to tell, that I got only a file with 0-length without any output.

Your distribution works differently to mine - can you please try `journalctl -f` as suggested by @solbu ?

I'm off to work shortly, so I'll be offline for a few hours...

**Comment by martin-ms** [comment]
> Your distribution works differently to mine - can you please try `journalctl -f` as suggested by @solbu ?

I did it yesterday after I read, that .xsession-errors is depreciated and logs are now written into the journal or Xorg.0.log, but I could not find any entries belonging to the weatherWidget; not as user and not as root. 

I don't know whats going wrong, to verify that I cought up with the right line & file, this is the diff:

```
diff -ur a/contents/ui/main.qml b/contents/ui/main.qml
--- a/contents/ui/main.qml	2022-02-15 00:28:38.000000000 +0100
+++ b/contents/ui/main.qml	2022-03-22 12:06:02.000000000 +0100
@@ -103,7 +103,7 @@
     Plasmoid.compactRepresentation: cr
     Plasmoid.fullRepresentation: fr
 
-    property bool debugLogging: plasmoid.configuration.debugLogging
+    property bool debugLogging: true
 
     FontLoader {
         source: '../fonts/weathericons-regular-webfont-2.0.10.ttf'
```

> I'm off to work shortly, so I'll be offline for a few hours...

Have a nice day & thanks to support us.

**Comment by cfint** [comment]
The user-agent is different with/without System Monitor widget in the tray.

Without System Monitor widget:  
```
  'HTTP_USER_AGENT': 'Mozilla/5.0 (X11; Linux x86_64) KIO/5.92 '
                     'plasmashell/5.24.3',
```

With System Monitor widget:
```
  'HTTP_USER_AGENT': 'Mozilla/5.0',
```


I used this bit of code to run a local http server to dump the headers. Code taken from [here](https://stackoverflow.com/a/25467602).
``` python
import os
import sys
import requests
from flask import Flask, json
from flask import Response

import pprint


class LoggingMiddleware(object):
    def __init__(self, app):
        self._app = app

    def __call__(self, env, resp):
        errorlog = env['wsgi.errors']
        pprint.pprint(('REQUEST', env), stream=errorlog)

        def log_response(status, headers, *args):
            pprint.pprint(('RESPONSE', status, headers), stream=errorlog)
            return resp(status, headers, *args)

        return self._app(env, log_response)

if __name__ == '__main__':
    app = Flask(__name__)
    app.wsgi_app = LoggingMiddleware(app.wsgi_app)
    app.run()
```

**Comment by blackadderkate** [comment]
> Request sent with User-Agent `-` on Arch. Response: `Please use a unique, identifying User-Agent header`
> 
> Details: [FAQ](https://api.met.no/doc/FAQ#nowimsuddenlygetting403forbiddenwhyamiblocked), [ToS](https://api.met.no/doc/TermsOfService#traffic) _Note: If you want to adhere to the rules in ToS, truncating coordinates should also be done._
> 
I have created issue #91 to truncate the co-ordinates...

> The user-agent is different with/without System Monitor widget in the tray.
> 
> Without System Monitor widget:
> 
> ```
>   'HTTP_USER_AGENT': 'Mozilla/5.0 (X11; Linux x86_64) KIO/5.92 '
>                      'plasmashell/5.24.3',
> ```
> 
> With System Monitor widget:
> 
> ```
>   'HTTP_USER_AGENT': 'Mozilla/5.0',
> ```
> 
> I used this bit of code to run a local http server to dump the headers. Code taken from [here](https://stackoverflow.com/a/25467602).
> 
> ```python
> import os
> import sys
> import requests
> from flask import Flask, json
> from flask import Response
> 
> import pprint
> 
> 
> class LoggingMiddleware(object):
>     def __init__(self, app):
>         self._app = app
> 
>     def __call__(self, env, resp):
>         errorlog = env['wsgi.errors']
>         pprint.pprint(('REQUEST', env), stream=errorlog)
> 
>         def log_response(status, headers, *args):
>             pprint.pprint(('RESPONSE', status, headers), stream=errorlog)
>             return resp(status, headers, *args)
> 
>         return self._app(env, log_response)
> 
> if __name__ == '__main__':
>     app = Flask(__name__)
>     app.wsgi_app = LoggingMiddleware(app.wsgi_app)
>     app.run()
> ```

That's very useful information, thanks. I will submit it to the KDE bug report I filed.

**Comment by blackadderkate** [comment]
Hi.
Just to let you know that I haven't forgotten about this bug - I've just started a new job and I haven't been able to spare any time maintaining this widget. When my training ends I should get a lot more free time to have a look at this issue further...

As far as I can tell, the bug in the system monitor widget has not been fixed yet... :(

**Comment by blackadderkate** [comment]
Latest update - according to [the KDE bug report](https://bugs.kde.org/show_bug.cgi?id=451790#c8), they patched the bug earlier today.
The fix will be released in KDE Framework version 5.95.

**Comment by fkru** [comment]
> Latest update - according to [the KDE bug report](https://bugs.kde.org/show_bug.cgi?id=451790#c8), they patched the bug earlier today. The fix will be released in KDE Framework version 5.95.

JFYI: KDE Frameworks 5.95 is expected to be released on 11 June. At least for openSUSE Tumbleweed it will be available a couple of days later.

**Comment by martin-ms** [comment]
The problem has been fixed upstream with the release 5.95.0 of KDE Frameworks and it works. No further actions are needed by the developer of the WeatherWidget-2.
